axelrod/strategies/axelrod_second.py

```diff
@@ -199,3 +202,210 @@ def strategy(self, opponent: Player) -> Action:
+
+        if self.num_turns_after_good_defection == 2:
+            self.num_turns_after_good_defection = 0
+            self.two_turns_after_good_defection_ratio = (
+                ((self.two_turns_after_good_defection_ratio 
+                * self.two_turns_after_good_defection_ratio_count) 
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - ((self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]]))) 
+                / (self.two_turns_after_good_defection_ratio_count + 1)
+                )
+            self.two_turns_after_good_defection_ratio_count += 1
+        elif self.num_turns_after_good_defection == 1:
+            self.num_turns_after_good_defection = 2
+            self.one_turn_after_good_defection_ratio = (
+                ((self.one_turn_after_good_defection_ratio 
+                * self.one_turn_after_good_defection_ratio_count)
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - (self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]])) 
+                / (self.one_turn_after_good_defection_ratio_count + 1)
+                )
+            self.one_turn_after_good_defection_ratio_count += 1
+   
+    def update_stochastic_state(self, opponent):
+            
+        if (self.current_score[0] / ((len(self.history)) + 1)) >= 2.25:
+                self.probability = (
+                    (.95 - (((self.one_turn_after_good_defection_ratio)
+                    + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
+                    + (1 / (((len(self.history))+1) ** 2))
+                    - (self.dict[opponent.history[-1]] / 4)
+                    )
+                self.probability = round(self.probability, 4)
+                self.score = "good"
+        elif (self.current_score[0] / ((len(self.history)) + 1)) >= 1.75:
+                self.probability = (
+                    (.25 + ((opponent.cooperations + 1) / ((len(self.history)) + 1)))
+                    - (self.opponent_consecutive_defections * .25) 
+                    + ((self.current_score[0] 
+                    - self.current_score[1]) / 100) 
+                    + (4 / ((len(self.history)) + 1))
+                    )
+                self.probability = round(self.probability, 4)
+                self.score = "average"
+
+    def strategy(self, opponent: Player) -> Action:
+
+        self.current_score = compute_final_score(zip(self.history, opponent.history))
+        
+        if len(self.history) == 0:
+            return C
+        else: 
+            self.update_cooperative_state(opponent)
+            if opponent.history[-1] == D: 
+                self.opponent_consecutive_defections += 1
+            else:
+                self.opponent_consecutive_defections = 0
+            self.update_stochastic_state(opponent)
```

Comment:
We can get rid of `update_stochastic_state` completely I think and just have the logic for the stochastic state directly here. The rest of the strategy then becomes (please check carefully):

```
if (self.current_score[0] / ((len(self.history)) + 1)) >= 2.25:
                probability = (
                    (.95 - (((self.one_turn_after_good_defection_ratio)
                    + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
                    + (1 / (((len(self.history))+1) ** 2))
                    - (self.dict[opponent.history[-1]] / 4)
                    )
               if random.random() <= probability:  # I have plans for changing this a bit but let's take that one step at a time
                   return C
               self.num_turns_after_good_defection = 1
               return D
if (self.current_score[0] / ((len(self.history)) + 1)) >= 1.75:
                probability = (
                   (.25 + ((opponent.cooperations + 1) / ((len(self.history)) + 1)))
                   - (self.opponent_consecutive_defections * .25) 
                   + ((self.current_score[0] 
                   - self.current_score[1]) / 100) 
                   + (4 / ((len(self.history)) + 1))
                   )
               if random.random() <= probability:
                   return C
               self.num_turns_after_good_defection = 1
               return D
return opponent.history[-1]
```

Reply:
Note that this in turn gets rid of a few of the internal variables (`self.probability` and `self.score`) you should change the `init` method to reflect that.